### PR TITLE
Hook in GI (probes) to SpeedTree.

### PR DESCRIPTION
**Pull request by russoh**
### **Please read**
**PR workflow guidelines**
* SRP ABV will start automatically on Yamato when you open your PR
* Changes to docs and md files will **not** trigger ABV jobs 
* Consider making use of **draft PRs** if you are not 100% sure that your PR is ready for review
* ABV will restart if you add a new commit to a branch with an open PR (hence why you should consider using draft PRs)
* Adding [skip ci] (case insensitive) to the title of PRs will stop any jobs being trigger automatically - you will need to open Yamato and find your branch to run ABV
* You can also add [skip ci] to commit messages to prevent CI from running on that push
* Add [cancel old ci] to your commit message if you've made changes you want to test and no longer need the previous jobs

### Checklist for PR maker
- [ ] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [ ] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [ ] Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR.
- [ ] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
Adding GI to SpeedTree (probes)

---
### Testing status

**Manual Tests**: What did you do?
Have updated and tested in existing graphics scenes for SpeedTree:

057_SpeedTree_V7
058_SpeedTree_V8

Have tested using lightmapping vs probes vs realtime. UPDATE: lightmapping removed as active variant.

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

Probes and light mapping tests have been added to:
058_SpeedTree_V8

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics

Any test projects to go with this to help reviewers?

https://fogbugz.unity3d.com/f/cases/1171489
Case_1171489.zip

You can validate in this project with an updated manifest that uses latest of Unity trunk & Graphics package.
E.g.:
    "com.unity.render-pipelines.core": "file:../../Graphics/com.unity.render-pipelines.core",
    "com.unity.render-pipelines.universal": "file:../../Graphics/com.unity.render-pipelines.universal",
    "com.unity.shadergraph": "file:../../Graphics/com.unity.shadergraph",
---
### Comments to reviewers
Tests need to be updated, and it will need backport labels TBD.

### CODE REVIEW

**Comment by LandonTownsendUnity** [comment]
The problems with the speedtree 8 shaders don't seem to be a problem with GI, they exist when GI is removed.

File: com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8Passes.hlsl
Comment (ellioman) [inline com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8Passes.hlsl:299]: This lines is giving errors. I assume it's because it needs to be `SAMPLE_GI(input.interpolated. interpolated.....)`

Comment (russoh) [reply com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8Passes.hlsl]: Good catch, looks like the test setup did not actually turn on the GI, so the code wasn't even getting compiled.

**Comment by nerites** [comment]
I wanna check some things about those tests before I update the reference images... there seem to be diffs in the AO on the bears, not just on the trees. Further, two of the trees are significantly different from before, and only one of those trees actually has GI enabled.

**Comment by nerites** [comment]
Ah okay, the trees are brighter since they are now sampling from SH even if _LIGHTMAP_ON isn't set. 
I'm still not sure about the AO differences on the bears, but maybe they alone aren't enough to cause a mismatch.

**Comment by nerites** [comment]
Queued up the iPhone and Android tests for universal as well.

What about Universal_Hybrid?

**Comment by ellioman** [comment]
> What about Universal_Hybrid?
Those are currently broken at the moment. I'm afraid we can't do much for those until that gets fixed.

> I'm still not sure about the AO differences on the bears, but maybe they alone aren't enough to cause a mismatch.
Yeah, I'm starting to worry that testing AO in this CI might be difficult. But do update the reference images with the new ones and let's see if they pass. I can follow up on this if it starts failing because of the AO itself.

**Comment by nerites** [comment]
Weird new wrinkle- for iPhone_Metal_Standalone, tests 57, 58, and 143 fail as expected, but 142 passes without any update to the reference image. S:

**Comment by phi-lira** [comment]
@russoh @LandonTownsendUnity can I merge this PR? I'm not sure if the issues reported on the discussion here are all fixed or not.

**Comment by nerites** [comment]
Of the failing tests, I confirmed that HDRP on Win_DX12_playmode and HDRP on Win_Vulkan_playmode pass with a custom editor revision that includes a fix for the issue.

I sitll can't get HDRP on OSX_Metal_playmode to pass; it keeps failing at the following point: 
[59/176] TestStart 'HDRP_GraphicTestRunner.2204_ReflectionProbes_Lights'

[19:27:26.519 INF] packet_write_wait: Connection to 172.28.218.33 port 22: Broken pipe

Since master HDRP is passing all builds, I'm going to pull master and hope that fixes it.

**Comment by nerites** [comment]
The VFX failure passed on rerun. I posted about the OSX_Metal failure in devs-yamato: https://unity.slack.com/archives/C998PJJDD/p1594856965140000

**Comment by phi-lira** [comment]
@nerites I think this should be safe, there are no changes in SRP Core or HDRP in this PR so I'm not sure how this PR could break those tests. Maybe it's an instability?

**Comment by phi-lira** [comment]
Q: How does speedtree work with lightmap? Isn't it suppose to only getting GI from sky probe/SH?

**Comment by nerites** [comment]
Is there a reason it shouldn't sample the lightmap? I don't know much about speedtree myself, but I don't see any reason it shouldn't... 

Yao mentioned that the GI texture coordinates from the speedtree assets didn't look right when he was working with them, is that what you mean? If so, do you want to just drop the LIGHTMAP_ON multicompile? 

I think I'd maybe wanna check with Russ when he gets back if we're doing that.

**Comment by phi-lira** [comment]
I'm not sure as well, thus why I ask. I know in theory there's a hack for procedural mesh renderers and LOD to receive lightmap data by, but I'm not sure about speedtree. It would be nice to confirm it works otherwise we pay cost of extra shader variants without properly supporting the feature.

**Comment by nerites** [comment]
After some research, Universal's speedtree implementation is the first one with GI hooked up. I was able to confirm that LIGHTMAP_ON gets set as expected and makes *some* sort of difference in existing Universal tests, but I'm unsure that the lightmapUVs are set up correctly.

I see that other Universal shaders call the following in the vertex shader:
    OUTPUT_LIGHTMAP_UV(input.lightmapUV, unity_LightmapST, output.lightmapUV);
    OUTPUT_SH(output.normalWS.xyz, output.vertexSH);

@russoh Do we need to add something like this to the speedtree shaders?

**Comment by nerites** [comment]
I tried adding those lines into SpeedTree7 and it looks a little funky. I set lightmapUV to use TEXCOORD4 for speedtree, but lightmapUV uses TEXCOORD1 everywhere else. For speedtree, TEXCOORD1 seems to be used for some position calculations. I'm not clear on how inputs to vertex shaders are populated- is this a problem?

**Comment by russoh** [comment]
It looks like manual visual testing wasn't enough to validate this correctly. I would suggest at this point that we need to fully validate the lightmap uv code path:
1. In engine code in trunk, check that uv streams are correctly setup for trees.
2. Double check the use of macros to declare, and propagate uvs to fragment shader.
3. Expand the automated tests , and have them manually verified. 
If a trunk change is actually required for the light map uvs, maybe we just support probes for now.

**Comment by nerites** [comment]
Connected SHs for now to see if that affects reference images.

**Comment by russoh** [comment]
SpeedTreeImporter.cpp, line 1929:
                    // lightmap uv
                    texCoord1.x = Vertex.m_vLightmapTexCoord.x;
                    texCoord1.y = Vertex.m_vLightmapTexCoord.y;
line 1993:
                    mesh->SetUv(1, reinterpret_cast<const float*>(&texCoords1[0]), 4, numVertices);

This is reading generated lightmap uvs from a SpeedTree 8 asset (exported using their tool). It looks like it gets bound to the correct UV set in the mesh. 

So it isn't correct in any circumstance for SpeedTree 7, so we definitely should disable lightmapping for SpeedTree 7, but I have it functioning correctly on SpeedTree 8, and will push that soon with the correct binding.

**Comment by russoh** [comment]
The UV islands looks sane when I check via the Lighting window, and the indirect contribution looks correct.

**Comment by russoh** [comment]
I re-enabled light-mapping for SpeedTree 8, after I checked the engine hooks, the shader macros, and verified the UV islands via the lighting window, and checked visual output with an updated test. It looks correct as far as I can tell. I'm not really sure what we do (if at all) for adding a light-mapping test given the bake time needed for it. If we want to add this to automation, might need some more insight on setup needed. In any case, I made a branch that just has a modified version of the SpeedTree 8 test scene, where I added extra trees with light-mapping and probes, and appropriate lights, and settings. I'm not sure if we can roll this back into the master. Maybe we can, it's here:

https://github.com/Unity-Technologies/Graphics/tree/universal/speedtreegi_updatedtest
058_SpeedTree_V8

**Comment by phi-lira** [comment]
What's the difference between Speedtree v7 and v8? How does the user select one or another?
@russoh we do have some light baking tests already, small scenes are usually fine for bake times. Is the time to bake terrain/speedtree significantly higher than other renderers?

**Comment by russoh** [comment]
> What's the difference between Speedtree v7 and v8? How does the user select one or another?
> @russoh we do have some light baking tests already, small scenes are usually fine for bake times. Is the time to bake terrain/speedtree significantly higher than other renderers?

It's the version of the SpeedTree asset, as authored and exported by the SpeedTree creation tool/editor. It's a proprietary format, with different vertex streams according to the version features. Version 8 has a lightmap UV stream in the assset, version 7 doesn't. There is separate import code for the 2 versions. For SpeedTree 8, the lightmap UV can be bound to mesh UV index 1 (uv2). For SpeedTree 7 it does need to be disabled because it's not provided in the imported data.

It was taking a notable amount of time to bake the lightmaps, but I'm not well acquainted with tuning the bake parameters, or sensible configuring all the lighting params, so perhaps it be be properly optimized by somebody with more knowledge (e.g. an artist).

**Comment by russoh** [comment]
The SpeedTree 8 test scene now includes GI.

**Comment by russoh** [comment]
I think all the issues are resolved at this point:
-GI probes for SpeedTree 7 
-GI lightmaps and probes for SpeedTree 8 (since we do have lightmap uvs)
-Updated test scene which has light map baking for SpeedTree 8

Will just need to follow up and make sure that the reference images are appropriate for all platforms.

File: com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8Passes.hlsl
Comment (KEngelstoft) [reply com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8Passes.hlsl]: did you add test coverage for this code path?

File: com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree7Billboard.shader
Comment (KEngelstoft) [inline com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree7Billboard.shader:36]: Lightmapping a billboard doesn't make sense, remove this.

File: TestProjects/UniversalGraphicsTest/Assets/Scenes/057_SpeedTree_V7.unity
Comment (KEngelstoft) [inline TestProjects/UniversalGraphicsTest/Assets/Scenes/057_SpeedTree_V7.unity:102]: You added baked lighting to the 058_SpeedTree_V8 test scene. Do you have a scene without baked lighting to make sure that other code path keep working?

**Comment by KEngelstoft** [comment]
Why are speedtrees placed in the 142_SSAO_DepthNormal_Projection test scene? Seems unrelated to what the test is testing...

**Comment by KEngelstoft** [comment]
Why are there black artifacts on the leftmost tree in 057_SpeedTree_V7?

File: com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8.shader
Comment (KEngelstoft) [inline com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8.shader:54]: It doesn't make much sense to lightmap trees, because they have a huge surface area, complicated unwrap with lots of charts so each tree will eat an entire lightmap atlas to get decent looking output. Light probes are a much better fit.

**Review by KEngelstoft — CHANGES_REQUESTED** [review]
Before merging this, we should have a conversation around the goal of this PR. Is it some kind of GI support for trees or GI via light probes AND _lightmapping_? Please reconsider if lightmap support is needed at all, as it doesn't have much practical usage because of the huge surface area of most tree assets. Light probes (SH L2) is a much better fit.
If we do want to support lightmaps, this needs investigation: some of the template images make me think that the UV stream is getting corrupted because it is both used to carry Speedtree data and lightmap UV data. This is done by following the data back to the speedtree importer and verifying that the data isn't overwritten/trashed by the lightmapping. gfx tests are not great for spotting problems like this.

**Comment by russoh** [comment]
> Why are speedtrees placed in the 142_SSAO_DepthNormal_Projection test scene? Seems unrelated to what the test is testing...

I asked the same thing, I don't know why trees are added to those tests, it seems like an undesirable test design. Somebody familiar with the SSAO test scene would remove that dependency next time the scene is updated.

File: com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8.shader
Comment (russoh) [reply com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8.shader:54]: > Before merging this, we should have a conversation around the goal of this PR. Is it some kind of GI support for trees or GI via light probes AND _lightmapping_? Please reconsider if lightmap support is needed at all, as it doesn't have much practical usage because of the huge surface area of most tree assets. Light probes (SH L2) is a much better fit.
> If we do want to support lightmaps, this needs investigation: some of the template images make me think that the UV stream is getting corrupted because it is both used to carry Speedtree data and lightmap UV data. This is done by following the data back to the speedtree importer and verifying that the data isn't overwritten/trashed by the lightmapping. gfx tests are not great for spotting problems like this.

I'm definitely not intimately knowledgeable of how lighting team have integrated probes and light-mapping, and potential implications for corrupted uv streams. It's perfectly reasonable for you guys to make the call to remove light mapping if that's the appropriate solution based on your knowledge of the functionality and performance implications versus probes etc. Confirm to remove?

File: TestProjects/UniversalGraphicsTest/Assets/Scenes/057_SpeedTree_V7.unity
Comment (russoh) [reply TestProjects/UniversalGraphicsTest/Assets/Scenes/057_SpeedTree_V7.unity:102]: > Why are there black artifacts on the leftmost tree in 057_SpeedTree_V7?

I think the scene is OK. The ref image looks darker around some leaves versus running the test scene in latest, looks like it just didn't trigger a new image output, presumably because it's within some error tolerance. We can force update it.

Comment (russoh) [reply TestProjects/UniversalGraphicsTest/Assets/Scenes/057_SpeedTree_V7.unity:102]: > You added baked lighting to the 058_SpeedTree_V8 test scene. Do you have a scene without baked lighting to make sure that other code path keep working?

There are also non-baked lights in this scene on other trees.

File: com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree7Billboard.shader
Comment (russoh) [reply com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree7Billboard.shader:36]: Right, that should have been removed.

**Comment by ellioman** [comment]
> > Why are speedtrees placed in the 142_SSAO_DepthNormal_Projection test scene? Seems unrelated to what the test is testing...
> 
> I asked the same thing, I don't know why trees are added to those tests, it seems like an undesirable test design. Somebody familiar with the SSAO test scene would ideally remove that dependency next time the scene is updated.

In my SSAO test scenes I wanted to cover all of our main shaders Lit, SimpleLit, BakedLit, SpeedTree 7 & 8 as well as some ShaderGraph shaders. 142 is testing then both that the DepthNormals pass is correct (although I should really make a separate test just for that) and then SSAO on top of that. 143 is then testing the same thing but with an orthographic camera.

File: com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8Passes.hlsl
Comment (KEngelstoft) [reply com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8Passes.hlsl]: Still waiting for a reply...

File: com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8Passes.hlsl
Comment (KEngelstoft) [inline com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8Passes.hlsl:271]: OUTPUT_LIGHTMAP_UV(input.texcoord1.xy, unity_LightmapST, output.lightmapUV); should not be needed if you have removed all traces of lightmapping. SH below should suffice.

**Review by KEngelstoft — APPROVED** [review]
I will go ahead and approve, but on condition that the automation I asked about here https://github.com/Unity-Technologies/Graphics/pull/720#discussion_r482880816 is there…

**Comment by GrantLamb-Unity** [comment]
I've created a JIRA task on the Terrain board in order to review @KEngelstoft's specific request regarding the automated test case once @russoh returns from vacation. https://jira.unity3d.com/browse/GFXTR-189

Could we get this HUP bugfix merged?

**Comment by phi-lira** [comment]
Merging given the ask from Kasper is tracked on JIRA and should be addressed in follow up PR.

**Pull request merged**
